prov/gni/src/gnix_mr.c

```diff
@@ -574,8 +644,8 @@ static int __mr_cache_register(
 	gni_return_t grc = GNI_RC_SUCCESS;
 
 	/* if we shouldn't introduce any new elements, return -FI_ENOSPC */
-	if (cache->attr.hard_reg_limit > 0 &&
-			atomic_get(&cache->inuse_elements) >= cache->attr.hard_reg_limit)
+	if (unlikely(cache->attr.hard_reg_limit > 0 &&
+			atomic_get(&cache->inuse_elements) >= cache->attr.hard_reg_limit))
 		return FI_ENOSPC;
```

Comment:
This check can prevent use of an existing MR cache entry that satisfies the call when the cache is full.